Table support

ralsina opened the pull request:
Working implementation.

There seems to be a bug in markd about escaping characters inside `<code>` which prevents the "Escaped pipes" test from passing but it's unrelated to table support.

It passes the tests in the gfm suite related to tables except those that rely on a working escaping support.

~~There is *one* regression test failing, need to look into it.~~

I consider this finished :-)

zw963 commented:
Wow, you're so efficient!

My fork of [markd(missing.style branch)](https://github.com/zw963/markd/tree/missing.style) now based on @ralsina 's `table-support` branch now, and i change my old `aweful` table to use new markdown syntax, it works!

Check original markdown [here](https://github.com/crystal-china/website/blob/master/markdowns/for_advanced_rubyists/methods.md) and rendered page [here](https://crystal-china.org/docs/for_advanced_rubyists/methods)

Not as good as github rendered table (no border, and no interlaced background colors), but it's works!

![image](https://github.com/user-attachments/assets/261ac0da-fdc9-4c50-8a3c-dae6beafb831)

nobodywasishere commented on src/markd/rules/table.cr:
```diff
@@ -0,0 +1,116 @@
+      has_body = lines.size > 2
+      container.data["has_body"] = has_body
+
+      alignments = lines[1].strip.strip("|").split(/(?<!\\)\|/).map do |cell|
```
nitpick: would it be useful to put this regex in a const as we're using it in several places?

nobodywasishere commented on src/markd/parsers/block.cr:
```diff
@@ -122,7 +123,7 @@ module Markd::Parser
         unless @indented
           first_char = @line[@next_nonspace]?
-          unless first_char && (Rule::MAYBE_SPECIAL.includes?(first_char) || first_char.ascii_number?)
+          unless first_char && (Rule::MAYBE_SPECIAL.includes?(first_char) || first_char.ascii_number? || @line.match(/(?<!\\)\|/))
```
Tables should only be parsed when `@options.gfm` is true

ralsina replied:
Ok, that was actually pretty easy. Done in latest commit.

ralsina commented:
Indeed. Fixed in latest commit.

On Mon, Feb 24, 2025 at 11:47 AM Margret Riegert ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/markd/rules/table.cr
> <https://github.com/icyleaf/markd/pull/72#discussion_r1967794431>:
>
> > +      if lines.size < 2 || !lines[1].match(SEPARATOR_REGEX) ||
> +         row_sizes.size != 1
> +        # Not enough table or a broken table.
> +        # We need to convert it into a paragraph
> +        # I am fairly sure this is not supposed to work
> +        container.type = Node::Type::Paragraph
> +        # Patch the text to have the leading |s
> +        container.text = original_text
> +        return
> +      end
> +
> +      max_row_size = row_sizes[0]
> +      has_body = lines.size > 2
> +      container.data["has_body"] = has_body
> +
> +      alignments = lines[1].strip.strip("|").split(/(?<!\\)\|/).map do |cell|
>
> nitpick: would it be useful to put this regex in a const as we're using it
> in several places?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/icyleaf/markd/pull/72#pullrequestreview-2637379870>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAMK7KB5IIXNGRJB2JILL2RMWJTAVCNFSM6AAAAABXWHYYIOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZXGM3TSOBXGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ralsina commented:
This is slightly tricky because it means that I need to see the options
from the Table rule and I am not sure how, will give it a shot later.

On Mon, Feb 24, 2025 at 11:57 AM Margret Riegert ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/markd/parsers/block.cr
> <https://github.com/icyleaf/markd/pull/72#discussion_r1967814698>:
>
> > @@ -122,7 +123,7 @@ module Markd::Parser
>          # this is a little performance optimization
>          unless @indented
>            first_char = @***@***.***_nonspace]?
> -          unless first_char && (Rule::MAYBE_SPECIAL.includes?(first_char) || first_char.ascii_number?)
> +          unless first_char && (Rule::MAYBE_SPECIAL.includes?(first_char) || first_char.ascii_number? || @line.match(/(?<!\\)\|/))
>
> Tables should only be parsed when @options.gfm is true
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/icyleaf/markd/pull/72#pullrequestreview-2637416674>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAMK7U44YS7OM6CWIHI7D2RMXPHAVCNFSM6AAAAABXWHYYIOVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDMMZXGQYTMNRXGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

zw963 commented:
Updated to tested on the latest commit, looks good!

![image](https://github.com/user-attachments/assets/9a0a58f6-4876-4d7b-b463-4f6209c88edf)

icyleaf commented:
It is big move forward, ready to merge it?

ralsina commented:
I think so!

On Thu, Feb 27, 2025, 6:56 AM icyleaf ***@***.***> wrote:

> It is big move forward, ready to merge it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/icyleaf/markd/pull/72#issuecomment-2687445520>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAMK74H2YGMFISG7L3XI32R3OLHAVCNFSM6AAAAABXWHYYIOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBXGQ2DKNJSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: icyleaf]*icyleaf* left a comment (icyleaf/markd#72)
> <https://github.com/icyleaf/markd/pull/72#issuecomment-2687445520>
>
> It is big move forward, ready to merge it?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/icyleaf/markd/pull/72#issuecomment-2687445520>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAAMK74H2YGMFISG7L3XI32R3OLHAVCNFSM6AAAAABXWHYYIOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDMOBXGQ2DKNJSGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
